llvm/lib/Target/AArch64/AArch64SelectionDAGInfo.cpp

```diff
@@ -76,12 +76,74 @@ SDValue AArch64SelectionDAGInfo::EmitMOPS(AArch64ISD::NodeType SDOpcode,
   }
 }
 
+SDValue AArch64SelectionDAGInfo::EmitSpecializedLibcall(
+    SelectionDAG &DAG, const SDLoc &DL, SDValue Chain, SDValue Dst, SDValue Src,
+    SDValue Size, RTLIB::Libcall LC) const {
+  const AArch64Subtarget &STI =
+      DAG.getMachineFunction().getSubtarget<AArch64Subtarget>();
+  const AArch64TargetLowering *TLI = STI.getTargetLowering();
+  TargetLowering::ArgListTy Args;
+  TargetLowering::ArgListEntry Entry;
+  Entry.Ty = DAG.getDataLayout().getIntPtrType(*DAG.getContext());
+  Entry.Node = Dst;
+  Args.push_back(Entry);
+
+  enum { SME_MEMCPY = 0, SME_MEMMOVE, SME_MEMSET } SMELibcall;
+  switch (LC) {
+  case RTLIB::MEMCPY:
+    SMELibcall = SME_MEMCPY;
+    Entry.Node = Src;
+    Args.push_back(Entry);
+    break;
+  case RTLIB::MEMMOVE:
+    SMELibcall = SME_MEMMOVE;
+    Entry.Node = Src;
+    Args.push_back(Entry);
+    break;
+  case RTLIB::MEMSET:
+    SMELibcall = SME_MEMSET;
+    if (Src.getValueType().bitsGT(MVT::i32))
+      Src = DAG.getNode(ISD::TRUNCATE, DL, MVT::i32, Src);
```

Comment:
I'm not sure when this would be needed? I tried to write a test case which would require truncating Src, but found that I could only pass a source type of i8 to the memset intrinsic.
I might have missed something though, if there is a case where we would need to truncate Src please can you add a test for it?

Reply:
I think I have a testcase where I could see this oprtation is required:
```
void foo(size_t n)  __arm_streaming_compatible
{
  memset(dst, 2, n);
}

```
And here is intermidate representation:
```
define dso_local void @foo(i64 noundef %n) local_unnamed_addr #0 {
entry:
  tail call void @llvm.memset.p0.i64(ptr nonnull align 1 @dst, i8 2, i64 %n, i1 false)
  ret void
}

```
```
(gdb) p Src->dump()
t4: i8 = Constant<2>
$1 = void
(gdb) n
107	    Entry.Node = Src;
(gdb) p Src->dump()
t6: i32 = Constant<2>
$2 = void
```

Reply:
Thanks for sharing this test @dtemirbulatov.
I think in your example, Src still needs to be extended? The case I was trying to write was one where Src requires truncating, for example from i64 to i32.

Reply:
@kmclaughlin-arm , I was not able to produce truncating test somehow.  Is it ok to add zero-extend test for this?